### DISCUSSION TITLE
openapi.json generation lacks enum in union discriminator field when using pydantic 2.10

### DISCUSSION THREAD

**Discussion by sbarber-dsi** [fastapi/fastapi#12973] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class MyBaseModel(RootModel):
    name: str
    pass

class My2ndModel(MyBaseModel):
    type: Literal['My2ndModel']
    pass

class My3rdModel(MyBaseModel):
    type: Literal['My3rdModel']
    pass

# etc ...

class ChooseAModel(RootModel):
    root: Annotated[
        Union['MyRecursiveModel', My2ndModel, My3rdModel, My4thModel],
        Field(discriminator='type'),
    ]
    pass

class MyRecursiveModel(MyBaseModel):
    type: Literal['MyRecursiveModel']
    list[ChooseAModel] = []
    pass

# Generated with pydantic 2.9
"My2ndModel": {
        "properties": {
          "name": {
            "type": "string",
            "title": "Name"
          },
          "type": {
            "type": "string",
            "enum": [
              "My2ndModel"
            ],
            "title": "Type"
          },
        },
        "type": "object",
        "required": [
          "name",
          "type",
        ],
        "title": "My2ndModel"
      }

# Generated with pydantic 2.10 - missing "enum"
"My2ndModel": {
        "properties": {
          "name": {
            "type": "string",
            "title": "Name"
          },
          "type": {
            "type": "string",
            "title": "Type"
          },
        },
        "type": "object",
        "required": [
          "name",
          "type",
        ],
        "title": "My2ndModel"
      }
```

### Description

With pydantic 2.9 we get the openapi.json we expect
With pydantic 2.10 the openapi.json is missing enum entries on discriminator properties.

### Operating System

Linux

### Operating System Details

Ubuntu 22.04

### FastAPI Version

0.115.5

### Pydantic Version

2.10.1

### Python Version

Python 3.10.12

### Additional Context

We're pinned to fastapi 0.111.1 but I pulled 0.115.5 and got the same result

**Comment by luzzodev** [comment, marked as answer]
Hello @sbarber-dsi while I can confirm your issue, I can add more details.
This is not directly addressed to Fastapi but comes from pydantic. 
This change has been introduced in v2.10.0 and is just reverting the behaviour of version before 2.7.

So if you want to keep your expected behaviour you need to pin pydantic >= 2.7, <= 2.9.

**Reply by sbarber-dsi** [reply]
Understood; is it not possible for fastapi to produce an openapi.json with enum names with pydantic 2.10?
Or is there some additional annotation we are supposed to do with 2.10?

**Reply by luzzodev** [reply]
According to [here](https://github.com/pydantic/pydantic/pull/10692) const and enum is not openapi compliant.
The spec is created from pydantic, and for now the only way for a literal to create an enum is to have multiple value
